fix(keycloak): update more than 10 sub-groups (#9690)

FabienSalles opened the pull request:
##### SUMMARY

Fixes #9690 

When we want to create or update a sub-group, we fetch children of the parent group to know which sub group exist and should be updated.
The [API is paginated by default with a maximum of 10 result](https://www.keycloak.org/docs-api/latest/rest-api/index.html#_get_adminrealmsrealmgroupsgroup_idchildren) and [the api call did not set the corresponding paremeter (`max`)](https://github.com/ansible-collections/community.general/blob/main/plugins/module_utils/identity/keycloak/keycloak.py#L1552).

This PR set `max` parameter with the number of existing sub-groups retrieve by another API endpoint

##### ISSUE TYPE
Bugfix Pull Request

##### COMPONENT NAME
community.general.keycloak_group

FabienSalles commented:
This PR fix the issue on my side but I couldn't launch integration tests.
I encountered this error : 


```
➜ ansible-test integration keycloak_group --docker default -v --allow-unsupported
Configured locale: en_US.UTF-8
Falling back to tests in "tests/integration/targets/" because "roles/test/" was not found.
Run command: docker -v
Detected "docker" container runtime version: Docker version 27.3.1, build ce12230
Run command: docker info --format '{{ json . }}'
Run command: docker version --format '{{ json . }}'
Container runtime: docker client=27.3.1 server=27.3.1 cgroup=v2
Assuming Docker is available on localhost.
Run command: docker image inspect quay.io/ansible/default-test-container:8.12.0
Run command: docker image inspect quay.io/ansible/ansible-test-utility-container:2.0.0
Run command: docker run --volume /sys/fs/cgroup:/probe:ro --name ansible-test-probe-w05jUWs4 --rm quay.io/ansible/a ...
Container host audit status: EPERM (-1)
Container host max open files: 1048576
Container loginuid: 4294967295 (not set)
Starting new "ansible-test-controller-w05jUWs4" container.
Run command: docker run --tmpfs /tmp:exec --tmpfs /run:exec --tmpfs /run/lock --volume /var/run/docker.sock:/var/ru ...
Adding "ansible-test-controller-w05jUWs4" to container database.
Run command: docker container inspect ansible-test-controller-w05jUWs4
Run command: docker run --pid host --privileged --name ansible-test-init-controller-w05jUWs4 --rm quay.io/ansible/a ...
Stream command with data: docker exec -i ansible-test-controller-w05jUWs4 /bin/sh
Scanning collection root: /home/fsalles/projects/github/ansible_collections
Including collection: community.general (3379 files)
Creating a payload archive containing 4398 files...
Created a 4420857 byte payload archive containing 4398 files in 0 seconds.
Run command with stdin: docker exec -i ansible-test-controller-w05jUWs4 tar oxzf - -C /root
Creating container database.
Stream command: docker exec ansible-test-controller-w05jUWs4 /usr/bin/env ANSIBLE_TEST_CONTENT_ROOT=/root/ansible_c ...
Configured locale: en_US.UTF-8
Falling back to tests in "tests/integration/targets/" because "roles/test/" was not found.
Parsing container database.
Run command: /usr/bin/python3.12 -c 'import cryptography'
Installing requirements for Python 3.12 (controller)
Stream command with data: /usr/bin/python3.12
Execute command: /usr/bin/python3.12 /tmp/ansible-test-myda__97-pip.py install --disable-pip-version-check -r requirements/ansible.txt -c requirements/constraints.txt
Run command: /usr/bin/python3.12 /root/ansible/test/lib/ansible_test/_util/target/tools/yamlcheck.py
Configuring target inventory.
Running keycloak_group integration test role
Initializing "/tmp/ansible-test-j6twz7rb-injector" as the temporary injector directory.
Injecting "/tmp/python-2txgnooi-ansible/python" as a execv wrapper for the "/usr/bin/python3.12" interpreter.
Stream command: ansible-playbook keycloak_group-pf3atz77.yml -i inventory -v
Using /root/ansible_collections/community/general/tests/output/.tmp/integration/keycloak_group-n7qfvbhl-ÅÑŚÌβŁÈ/tests/integration/integration.cfg as config file
running playbook inside collection community.general
ERROR! couldn't resolve module/action 'community.docker.docker_container'. This often indicates a misspelling, missing collection, or incorrect module path.

The error appears to be in '/root/ansible_collections/community/general/tests/output/.tmp/integration/keycloak_group-n7qfvbhl-ÅÑŚÌβŁÈ/tests/integration/targets/keycloak_group/tasks/main.yml': line 6, column 3, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:


- name: Start container
  ^ here
NOTICE: To resume at this test target, use the option: --start-at keycloak_group
FATAL: Command "ansible-playbook keycloak_group-pf3atz77.yml -i inventory -v" returned exit status 4.
Run command with stdout: docker exec -i ansible-test-controller-w05jUWs4 sh -c 'tar cf - -C /root/ansible_collectio ...
Run command with stdin: tar oxzf - -C /home/fsalles/projects/github/ansible_collections/community/general/tests
FATAL: Command "docker exec ansible-test-controller-w05jUWs4 /usr/bin/env ANSIBLE_TEST_CONTENT_ROOT=/root/ansible_collections/community/general LC_ALL=en_US.UTF-8 /usr/bin/python3.12 /root/ansible/bin/ansible-test integration keycloak_group -v --allow-destructive --containers '{}' --allow-unsupported --truncate 120 --color yes --host-path tests/output/.tmp/host-w52xrjiy --metadata tests/output/.tmp/metadata-dgufm4ea.json" returned exit status 1.
Run command: docker stop --time 0 ansible-test-controller-w05jUWs4
Run command: docker rm ansible-test-controller-w05jUWs4
```

felixfontein commented:
Thanks for your contribution!

You need the community.docker collection to run the keycloak integration tests (see `ERROR! couldn't resolve module/action 'community.docker.docker_container'`).

FabienSalles commented:
Thanks for your reply.

How I can install it correctly for ansible integration tests ?

It is already installed on my computer but the error still persist : 
```
general on 🐳 v27.3.1 (docker.sock) …
➜ ansible-galaxy collection install community.docker                                                                
Starting galaxy collection install process
Nothing to do. All requested collections are already installed. If you want to reinstall them, consider using `--force`.

general on 🐳 v27.3.1 (docker.sock) …
➜ ansible-galaxy collection install community.docker --force
Starting galaxy collection install process
Process install dependency map
Starting collection install process
Downloading https://galaxy.ansible.com/api/v3/plugin/ansible/content/published/collections/artifacts/community-docker-4.3.1.tar.gz to /home/fsalles/.ansible/tmp/ansible-local-375385_8pxmj8z/tmpkcuvmf0h/community-docker-4.3.1-ijzp0_g1
Installing 'community.docker:4.3.1' to '/home/fsalles/.ansible/collections/ansible_collections/community/docker'
community.docker:4.3.1 was installed successfully
'community.library_inventory_filtering_v1:1.0.0' is already installed, skipping.
```

felixfontein commented:
Ah, I missed that you're using `--docker`. Then you must have it installed in the same tree where the community.general checkout lives, i.e. if the c.g checkout is in `/path/to/ansible_collections/community/general`, community.docker needs to be in `/path/to/ansible_collections/community/docker`.

russoz commented:
Hi @FabienSalles 

when you pass `--docker default` in the command line, that means the entire test will execute within a docker container. Now as part of this test, as you have noticed, it tries to install docker then run a keycloak container, _inside the test container_. It uses `setup_docker` for that - which I wrote - and it's been a while but I am pretty sure I did not consider that scenario (running docker inside docker). It was rather designed to run docker from a VM instead.

If I may suggest, give it a try to `andebox` by yours truly:

https://pypi.org/project/andebox/

You can run this somewhat easily in your local machine using vagrant by simply running:

```
$ andebox vagrant -n ubuntu-noble -s -- --allow-unsupported keycloak_group
```
In fact I have just executed this (on the main branch, so not testing your PR at the moment) and other than the violent aggression to the swap partitions of my old laptop, everything ran just fine. ;-)

Naturally you will need vagrant installed and you will have to create your Vagrantfile at the project dir (do not commit that file). Mine looks like this for the ubuntu part, YMMV:
```ruby
  config.vm.define "ubuntu-noble" do |noble|   # Ubuntu 24.04
    noble.vm.hostname = 'ubuntu-noble'
    noble.vm.box = "jcpetro97/ubuntu2404"
    noble.vm.synced_folder ".", "/vagrant", rsync__exclude: ['.tox/', '.git/', '.vagrant']
    noble.vm.provider "virtualbox" do |vb|
      vb.customize ["modifyvm", :id, "--audio", "none"]
      vb.customize ["modifyvm", :id, "--memory", "4096"]
    end
    noble.vm.provision "shell", inline: <<-SHELL
      apt-get -y update
      apt-get -y install python3-pip virtualenv python3-venv
      virtualenv --python 3 /venv
      /venv/bin/pip install -U pip setuptools
      /venv/bin/pip install andebox ansible-core decorator
    SHELL
  end
```

FabienSalles commented:
@russoz thank you a lot for the explanation

I managed to get the tests to work but : 

- I'm forced to destroy the VM after each execution otherwise the tests crashed
- test execution is really long (4 min each time)
- and now, all tests passed even if I removed my fix and I added tests to highlight the bug (see below my tests)

To finish, I don't think integration tests are the best place to test this. 
There are already tests that call the corresponding API, I should create a unit test to verify that I am using and sending the parameter correctly in the api call, but my poor knowledge of python prevents me from doing it.

I'm sorry, I don't think I'll have time to go any further.
I don't know your criteria for merging the PR but the CI passed, I didn't break anything and my fix unblocked my client. 

Would it be possible to stop there?

Tanks again for your time (you to @felixfontein :pray:)

See below my Ansible integration tests which should fail (for the update case) but pass even if I remove my fix : 

```
- name: Create more than 10 sub groups using parent name
  community.general.keycloak_group:
    auth_keycloak_url: "{{ url }}"
    auth_realm: "{{ admin_realm }}"
    auth_username: "{{ admin_user }}"
    auth_password: "{{ admin_password }}"
    realm: "{{ realm }}"
    name: "subgrp{{ item }}"
    parents:
      - name: rootgrp
  loop: "{{ range(0, 12)|list }}"
  register: result

- name: Assert that subgroup was correctly created
  assert:
    that:
      - item is not failed
  loop: "{{ result.results }}"

- name: Update more than 10 sub groups using parent name
  community.general.keycloak_group:
    auth_keycloak_url: "{{ url }}"
    auth_realm: "{{ admin_realm }}"
    auth_username: "{{ admin_user }}"
    auth_password: "{{ admin_password }}"
    realm: "{{ realm }}"
    name: "subgrp{{ item }}"
    parents:
      - name: rootgrp
  loop: "{{ range(0, 12)|list }}"
  register: result

- name: Assert that subgroup was correctly updated
  assert:
    that:
      - item is not failed
  loop: "{{ result.results }}"
```

russoz commented:
Hi @FabienSalles if you don't mind, keep the PR open (and don't delete the branch), we can take this slowly - no rush from our side. About your initial comments:
- Need to destroy the VM: 99.9% of the times, that derives from tests that do not clean up after themselves. In this case I find it a bit weird because, in theory, keycloak is running _inside a container, inside the VM_, so if we kill the container and run the test again, it should be a new container. Maybe that is all that is missing, but I would need to verify that.
- Test is long: well, need to break the eggs to make the omelette. A fresh VM running this test spends time: a) coming up, b) installing and configuring docker, then finally c) running the tests. If we can solve the previous item (VM restart) this is bound to go down, but nonetheless I believe it will still be somewhat long.
- As it comes down to your change and how/if tests passed then all I can say is that it happens to all of us :-) but let's fix these other issues and we can take a closer look into it (again, no rush).

How does that sound to you?

russoz commented:
I can confirm the tests are working in the current state of the branch. Will take a further look as time allows.

felixfontein commented:
Does this mean we can merge this once the PR has a [changelog fragment](https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-a-changelog-fragment)?

russoz commented:
> Does this mean we can merge this once the PR has a [changelog fragment](https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-a-changelog-fragment)?

Pretty much, yes.

russoz reviewed (approved):
LGTM

felixfontein commented on changelogs/fragments/9692-update-more-than-10-keycloak-sub-groups.yml:
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+    - keycloak - update more than 10 sub-groups (https://github.com/ansible-collections/community.general/pull/9692).
```
```suggestion
    - keycloak - update more than 10 sub-groups (https://github.com/ansible-collections/community.general/issues/9690, https://github.com/ansible-collections/community.general/pull/9692).
```
